##Installing PySpark

In [1]:
pip install pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 281.3 MB 49 kB/s 
     |████████████████████████████████| 199 kB 65.5 MB/s 
  Created wheel for pyspark: filename=pyspark-3.3.0-py2.py3-none-any.whl size=281764026 sha256=6a7ea8fe1bf148ebc51f4168236909952e39da0d56175a10b046bc3080f35083
  Stored in directory: /root/.cache/pip/wheels/7a/8e/1b/f73a52650d2e5f337708d9f6a1750d451a7349a867f928b885
Successfully built pyspark


##Importing Necessary Libraries

In [2]:
from pyspark import SparkConf
from pyspark.sql import SparkSession
from pyspark.sql.functions import sum, col, desc, asc, count
from pyspark.sql import functions as F
from pyspark.sql.functions import udf

In [3]:
spark = SparkSession.builder.appName("practice").getOrCreate()
spark.sparkContext.getConf().getAll();

In [5]:
df1 = spark.read.format("csv").option("header","true").option("inferSchema","true").load("/content/Airline_data.csv")# reading the file
df1.show()#its shows the 20 records


+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|Year|Month|DayofMonth|DayOfWeek|DepTime|CRSDepTime|ArrTime|CRSArrTime|UniqueCarrier|FlightNum|TailNum|ActualElapsedTime|CRSElapsedTime|AirTime|ArrDelay|DepDelay|Origin|Dest|Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|
+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|1989|    1|        23|        1|   1419|      1230|   1742|      1552|           UA|      183

In [ ]:
  row = df1.count() # total number of rows in dataframe
  # extracting number of columns from the Dataframe
  column = len(df1.columns) # total number of columns in dataframe
 

In [ ]:
 print(f'Number of Rows are: {row}')# printing the total rows
 print(f'Number of Columns are: {column}')#printing the total column

Number of Rows are: 426
Number of Columns are: 29


##1.Show a sample of 5 records from dataset.

In [ ]:
df1.limit(5).show()#5 records from the  dataset.

+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|Year|Month|DayofMonth|DayOfWeek|DepTime|CRSDepTime|ArrTime|CRSArrTime|UniqueCarrier|FlightNum|TailNum|ActualElapsedTime|CRSElapsedTime|AirTime|ArrDelay|DepDelay|Origin|Dest|Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|
+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|1989|    1|        23|        1|   1419|      1230|   1742|      1552|           UA|      183

 observation-
 * limit() is a DataFrame transformation that returns a DataFrame with top N rows

##2.Read the data with data types.

In [8]:
df1.show()#only showing top 20 rows
df1.printSchema() # its shows the datatypes

+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|Year|Month|DayofMonth|DayOfWeek|DepTime|CRSDepTime|ArrTime|CRSArrTime|UniqueCarrier|FlightNum|TailNum|ActualElapsedTime|CRSElapsedTime|AirTime|ArrDelay|DepDelay|Origin|Dest|Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|
+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|1989|    1|        23|        1|   1419|      1230|   1742|      1552|           UA|      183

observation-
* printSchema() is used to print or display the schema of the DataFrame in the tree format along with column name and data type
* show() is used to display the contents of the DataFrame in a Table Row & Column Format. By default it shows only 20 Rows and the column values are truncated at 20 characters.

##3.Make a new column MonthStr, Which has months in form of 01, 02, 03, ..., 12.

In [ ]:

def month(x):
  
     return("%02d" % (x))# %02d performs decimal integer conversion d, formatted with zero padding (0 flag), with width 2.
 


#month(1)

a=udf(month)#converting the udf assinging the variables a
df1 = df1.withColumn("MonthStr",a(col("Month")))#adding new column by using withColumn
df1.show()#its show the result

+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+--------+
|Year|Month|DayofMonth|DayOfWeek|DepTime|CRSDepTime|ArrTime|CRSArrTime|UniqueCarrier|FlightNum|TailNum|ActualElapsedTime|CRSElapsedTime|AirTime|ArrDelay|DepDelay|Origin|Dest|Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|MonthStr|
+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+--------+
|1989|    1|        23|        1|   1419|      1230|   1742|      1

In [ ]:
df1.select("MonthStr").show()# successfully add Monthstr column from the dataset

+--------+
|MonthStr|
+--------+
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
|      01|
+--------+
only showing top 20 rows



##4.Find the Number(#) of flights each airline made.

In [ ]:
df1.agg({"UniqueCarrier": "count"}).show()#using aggregate functions count ,counting the UniqueCarrier column

+--------------------+
|count(UniqueCarrier)|
+--------------------+
|                 426|
+--------------------+



observation-
* count()  Count the number of elements present in the dataframe 

* There are UniqueCarrier is 426 from the dataframe

##5.Find the mean Arrival Delay per origination airport. 

In [ ]:
delay_per_origination_airport=df1.groupBy("Origin").agg({"ArrDelay":"avg"})# usimg groupby from origin column and counting the arrdelay column ,after the signning the delay_per_origination_airport
delay_per_origination_airport.show()#its show the result

+------+-------------------+
|Origin|      avg(ArrDelay)|
+------+-------------------+
|   LIH|0.16666666666666666|
|   HNL|  14.21774193548387|
|   EWR|               9.25|
|   DEN| 20.166666666666668|
|   IAD| 12.966666666666667|
|   SFO| 11.215384615384615|
|   PHL|  6.827586206896552|
|   OGG|  16.24137931034483|
+------+-------------------+



Observation-
 * groupBy() function is used to collect the identical data into groups on DataFrame and perform aggregate functions on the grouped data.
 * avg()  is used to return the average value from a particular column in the DataFrame.

Output Explanation-

* Lih  has a 0.16666666666666666 arrdelay
* HNL has  a 14.21774193548387  arrdelay
* EWR has  a9.24 arrdelay
* DEN has a 20.166666666666668  arrdelay
* IAD has a 12.966666666666667 arrdelay
*   SFO has a11.215384615384615 arrdelay
* PHL|  6.827586206896552 arrdelay
* OGG|  16.24137931034483 arrdelay

##6.What is the average departure delay from each airport? 

In [ ]:
average_departure_delay=df1.groupBy("Origin").agg({"DepDelay":"avg"})# usimg groupby from origin column and counting the DepDelay column ,after the signning the average_departure_delay
average_departure_delay.show()#its show the result

+------+-------------------+
|Origin|      avg(DepDelay)|
+------+-------------------+
|   LIH|-3.7666666666666666|
|   HNL|  3.217741935483871|
|   EWR|  4.958333333333333|
|   DEN|               27.6|
|   IAD|                8.9|
|   SFO| 19.646153846153847|
|   PHL| 16.137931034482758|
|   OGG|                6.0|
+------+-------------------+






Observation-
 * groupBy() function is used to collect the identical data into groups on DataFrame and perform aggregate functions on the grouped data.
 * avg()  is used to return the average value from a particular column in the DataFrame.

Output Explanation-

*  LIH  has a 3.7666666666666666 DepDelay
*   HNL has a  3.217741935483871 DepDelay
*   EWR has a 4.958333333333333 DepDelay
*  DEN has a               27.6 DepDelay
*  IAD has a              8.9 DepDelay
*  SFO has a 19.646153846153847 DepDelay
* PHL has a 16.137931034482758 DepDelay
*  OGG     has a           6.0 DepDelay